### PR TITLE
Analysis Printer

### PR DESCRIPTION
**Pull request by fabianbs96**

### CODE REVIEW

**Comment by fabianbs96** [comment]
See [Coding Conventions](https://github.com/secure-software-engineering/phasar/wiki/Coding-Conventions)

File: include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h [hunk 1]
```diff
@@ -286,6 +288,8 @@ class IDEExtendedTaintAnalysis
   AbstractMemoryLocationFactory<d_t> FactFactory;
   const llvm::DataLayout &DL;
 
+  // TaintAnalysisPrinter *Printer{};
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h]: Remove the commented lines

File: include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h [hunk 2]
```diff
@@ -23,7 +23,9 @@
 #include "phasar/PhasarLLVM/Pointer/LLVMAliasInfo.h"
 #include "phasar/PhasarLLVM/TaintConfig/LLVMTaintConfig.h"
 #include "phasar/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.h"
+#include "phasar/PhasarLLVM/Utils/AnalysisPrinter.h"
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h]: do we really need this include?

File: include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h [hunk 3]
```diff
@@ -23,7 +23,9 @@
...
 #include "phasar/PhasarLLVM/Utils/BasicBlockOrdering.h"
+#include "phasar/PhasarLLVM/Utils/DataFlowAnalysisType.h"
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.h]: do we need this include?

File: include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.h [hunk 4]
```diff
@@ -13,6 +13,8 @@
 #include "phasar/DataFlow/IfdsIde/IDETabulationProblem.h"
 #include "phasar/PhasarLLVM/Domain/LLVMAnalysisDomain.h"
 #include "phasar/PhasarLLVM/Pointer/LLVMAliasInfo.h"
+#include "phasar/PhasarLLVM/Utils/AnalysisPrinter.h"
+#include "phasar/PhasarLLVM/Utils/DataFlowAnalysisType.h"
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.h]: Same as above? Do we need this includes?

File: include/phasar/PhasarLLVM/Utils/DataFlowAnalysisType.def [hunk 5]
```diff
@@ -27,6 +27,4 @@ DATA_FLOW_ANALYSIS_TYPES(IntraMonoFullConstantPropagation, "intra-mono-fca", "Si
 DATA_FLOW_ANALYSIS_TYPES(IntraMonoSolverTest, "intra-mono-solvertest", "Empty analysis. Just to see that the intraprocedural monotone solver works")
 DATA_FLOW_ANALYSIS_TYPES(InterMonoSolverTest, "inter-mono-solvertest", "Empty analysis. Just to see that the interprocedural monotone solver works")
 DATA_FLOW_ANALYSIS_TYPES(InterMonoTaintAnalysis, "inter-mono-taint", "Simple taint analysis using the monotone framework with k-limited call-strings. Use ifds-taint or ide-xtaint instead.")
-
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/Utils/DataFlowAnalysisType.def]: add these lines back, so this file is unchanged

File: lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDELinearConstantAnalysis.cpp [hunk 6]
```diff
@@ -12,7 +12,6 @@
 #include "phasar/DataFlow/IfdsIde/EdgeFunctionUtils.h"
 #include "phasar/DataFlow/IfdsIde/EdgeFunctions.h"
 #include "phasar/DataFlow/IfdsIde/FlowFunctions.h"
-#include "phasar/DataFlow/IfdsIde/IDETabulationProblem.h"
```
Comment (sritejakv) [inline lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDELinearConstantAnalysis.cpp]: why did you remove this include?

File: lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.cpp [hunk 7]
```diff
@@ -44,6 +47,12 @@ namespace psr {
 ///
 ///
 
+template <>
+std::string
+lToString<TypeStateDescription::State>(TypeStateDescription::State V) {
+  return std::to_string(V);
```
Comment (sritejakv) [inline lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.cpp]: you can use V.stateToString() function to get the string and also remove string include above

File: lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.cpp [hunk 8]
```diff
@@ -17,6 +17,8 @@
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/LLVMZeroValue.h"
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/TypeStateDescriptions/TypeStateDescription.h"
 #include "phasar/PhasarLLVM/Pointer/LLVMAliasInfo.h"
+#include "phasar/PhasarLLVM/Utils/AnalysisPrinter.h"
+#include "phasar/PhasarLLVM/Utils/DataFlowAnalysisType.h"
```
Comment (sritejakv) [inline lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDETypeStateAnalysis.cpp]: Same as above. do we need these includes here?

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    std::vector<int> KeysToRemove;
+    for (const auto &Entry : Map1) {
+      auto Iter = Map2.find(Entry.first);
+      if (Iter != Map2.end() && Iter->second == Entry.second) {
+        KeysToRemove.push_back(Entry.first);
+      }
+    }
+    for (const auto &Key : KeysToRemove) {
+      Map2.erase(Key);
+    }
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp:38]: Use iterator based for loop and remove the element during the iteration. so you can return early without loop over the whole container.

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+using CallBackPairTy = std::pair<IDEExtendedTaintAnalysis<>::config_callback_t,
+                                 IDEExtendedTaintAnalysis<>::config_callback_t>;
+
+// Use template to variate between Typesate and Taint analysis
+class GroundTruthCollector
+    : public AnalysisPrinter<IDEExtendedTaintAnalysisDomain> {
+public:
+  // constructor init Groundtruth in each fixture
+  GroundTruthCollector(map<int, set<string>> &GroundTruth)
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp]: Use llvm::DenseMap instead of std::map and llvm::StringSet instead of std::set (change all other occurances). llvm container have better performance optimizations.

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      : GroundTruth(GroundTruth){};
+
+  void findAndRemove(std::map<int, std::set<std::string>> &Map1,
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp]: same comment on std::map and std:.set as above

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+#include "phasar/PhasarLLVM/SimpleAnalysisConstructor.h"
+#include "phasar/PhasarLLVM/Utils/LLVMIRToSrc.h"
+#include "phasar/PhasarLLVM/Utils/LLVMShorthands.h"
+
+#include "TestConfig.h"
+#include "gtest/gtest.h"
+
+#include <nlohmann/json.hpp>
+
+using namespace std;
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp]: do include the while std namepsace in a cpp file. remove this line and only include the necessary standard library headers.

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        KeysToRemove.push_back(Entry.first);
+      }
+    }
+    for (const auto &Key : KeysToRemove) {
+      Map2.erase(Key);
+    }
+  }
+
+  void onResult(Warnings<IDEExtendedTaintAnalysisDomain> War) override {
+    map<int, set<string>> FoundLeak;
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp]: same comment on std::map and std:.set as above

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    FoundLeak.emplace(SinkId, LeakedValueIds);
+    findAndRemove(FoundLeak, GroundTruth);
+  }
+
+  void onFinalize(llvm::raw_ostream & /*OS*/ = llvm::outs()) const override {
+    EXPECT_TRUE(GroundTruth.empty());
+  }
+
+private:
+  map<int, set<string>> GroundTruth;
```
Comment (sritejakv) [inline unittests/Utils/AnalysisPrinterTest.cpp]: same comment on std::map and std:.set as above

File: lib/PhasarLLVM/Utils/LLVMSrcMgrAnalysisPrinter.cpp
Comment (sritejakv) [inline lib/PhasarLLVM/Utils/LLVMSrcMgrAnalysisPrinter.cpp:1]: Remove this empty file and its corresponding header file

File: lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.cpp [hunk 15]
```diff
@@ -13,12 +13,16 @@
 #include "phasar/DataFlow/IfdsIde/FlowFunctions.h"
 #include "phasar/DataFlow/IfdsIde/IDETabulationProblem.h"
 #include "phasar/PhasarLLVM/ControlFlow/LLVMBasedICFG.h"
+#include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/ExtendedTaintAnalysis/EdgeDomain.h"
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/ExtendedTaintAnalysis/GenEdgeFunction.h"
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/ExtendedTaintAnalysis/Helpers.h"
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/ExtendedTaintAnalysis/KillIfSanitizedEdgeFunction.h"
 #include "phasar/PhasarLLVM/DataFlow/IfdsIde/Problems/ExtendedTaintAnalysis/TransferEdgeFunction.h"
+#include "phasar/PhasarLLVM/Domain/LLVMAnalysisDomain.h"
```
Comment (sritejakv) [inline lib/PhasarLLVM/DataFlow/IfdsIde/Problems/IDEExtendedTaintAnalysis.cpp]: do we need this include?

File: include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+#include <optional>
+#include <type_traits>
+#include <vector>
+
+namespace psr {
+
+template <typename AnalysisDomainTy> struct Warnings {
+  typename AnalysisDomainTy::n_t Instr;
+  typename AnalysisDomainTy::d_t Fact;
+  typename AnalysisDomainTy::l_t LatticeElement;
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h]: Change these three lines to 

using n_t = typename AnalysisDomainTy::n_t;
using d_t = typename AnalysisDomainTy::d_t;
using l_t = typename AnalysisDomainTy::l_t;

and so you can just use n_t Inst, d_t DfFact, ... and is more readable

File: include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+      if constexpr (std::is_same_v<typename AnalysisDomainTy::l_t,
+                                   BinaryDomain>) {
+        OS << "Value: " << psr::lToString(Iter.LatticeElement) << "\n";
+      }
+    }
+  }
+
+private:
+  Results<AnalysisDomainTy> AnalysisResults;
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h]: Default initialize the member variable

`Results<AnalysisDomainTy> AnalysisResults{};`

File: include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+template <typename AnalysisDomainTy> struct Results {
+  std::vector<Warnings<AnalysisDomainTy>> War;
+};
+
+// TODO: explicit-template-instantiation
+template <typename AnalysisDomainTy> class AnalysisPrinter {
+public:
+  virtual ~AnalysisPrinter() = default;
+  AnalysisPrinter() : AnalysisResults{.War = {}} {}
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h]: If you default initialize the member variable, you can change this line to 

` AnalysisPrinter() = default;`

File: include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+      OS << "\tFact: " << psr::dToString(Iter.Fact) << "\n";
+
+      if constexpr (std::is_same_v<typename AnalysisDomainTy::l_t,
+                                   BinaryDomain>) {
+        OS << "Value: " << psr::lToString(Iter.LatticeElement) << "\n";
+      }
+    }
+  }
+
+private:
```
Comment (sritejakv) [inline include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h]: What if, I want to create a derived class of AnalysisPrinter without class variable. In the current version, we will still allocate memory to the class variable and not use it (similar to NullAnalysisPrinter). To overcome this, you could introduce another super class follows,

```
template <typename AnalysisDomainTy> class AnalysisPrinterBase {
  virtual void onResult(Warnings<AnalysisDomainTy> War) = 0;
  virtual void onInitialize() = 0;
  virtual void onFinalize(llvm::raw_ostream &OS = llvm::outs()) = 0;
}
```

and rename current AnalysisPrinter to DefaultAnalysisPrinter and inherit it from AnalysisPrinterBase as follows,

```
template <typename AnalysisDomainTy> class DefaultAnalysisPrinter: public AnalysisPrinter {
....
}
```

and so, NullAnalysisPrinter would be as follows,

```
template <typename AnalysisDomainTy>
class NullAnalysisPrinter final : public AnalysisPrinterBase<AnalysisDomainTy> {
...
}
```

**Review by sritejakv — DISMISSED** [review]
Looks good to me

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    }
+  }
+
+  void onResult(Warning<IDEExtendedTaintAnalysisDomain> War) override {
+    llvm::DenseMap<int, std::set<std::string>> FoundLeak;
+    int SinkId = stoi(getMetaDataID(War.Instr));
+    std::set<std::string> LeakedValueIds;
+    LeakedValueIds.insert(getMetaDataID((War.Fact)->base()));
+    FoundLeak.try_emplace(SinkId, LeakedValueIds);
+    findAndRemove(FoundLeak, GroundTruth);
```
Comment (fabianbs96) [inline unittests/Utils/AnalysisPrinterTest.cpp:47]: This looks overly complicated and should be improved in the upcoming ticket

File: unittests/Utils/AnalysisPrinterTest.cpp [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+TEST_F(AnalysisPrinterTest, XTaint01) {
+  llvm::DenseMap<int, std::set<std::string>> GroundTruth;
+
+  GroundTruth[15] = {"8"};
+  GroundTruthCollector GroundTruthPrinter = {GroundTruth};
+  doAnalysisTest("xtaint01_cpp.ll", GroundTruthPrinter, std::monostate{});
+}
+
```
Comment (fabianbs96) [inline unittests/Utils/AnalysisPrinterTest.cpp:123]: We should not just test this on only two test cases. There should be more. To be added in the upcoming ticket!
